README.md
# FS_Nano33BLE Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/FS_Nano33BLE.svg?)](https://www.ardu-badge.com/FS_Nano33BLE)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/FS_Nano33BLE.svg)](https://github.com/khoih-prog/FS_Nano33BLE/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/FS_Nano33BLE/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/FS_Nano33BLE.svg)](http://github.com/khoih-prog/FS_Nano33BLE/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents

* [Why do we need this FS_Nano33BLE library](#why-do-we-need-this-FS_Nano33BLE-library)
  * [Important Notes](#Important-Notes)
  * [Features](#features)
  * [Currently supported Boards](#currently-supported-boards)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Examples](#examples)
  * [  1. FS_Counting](examples/FS_Counting)
  * [  2. FS_Test](examples/FS_Test)
* [Example FS_Test](#example-fs_test)
* [Debug Terminal Output Samples](#debug-terminal-output-samples)
  * [1. FS_Counting on Nano 33 BLE with LittleFS size 256KB](#1-fs_counting-on-nano-33-ble-with-littlefs-size-256kb)
  * [2. FS_Test on Nano 33 BLE with LittleFS size 256KB](#2-fs_test-on-nano-33-ble-with-littlefs-size-256kb)
  * [3. FS_Test on Nano 33 BLE with FATFS size 512KB](#3-fs_test-on-nano-33-ble-with-fatfs-size-512kb)
* [Debug](#debug)
* [Troubleshooting](#troubleshooting)
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [License](#license)
* [Copyright](#copyright)

---
---

### Why do we need this [FS_Nano33BLE library](https://github.com/khoih-prog/FS_Nano33BLE)

## Important Notes

Avoid using `FATFS` because the somehow (issue with the core ???) it's OK to use only with `512KB`. Please use the better `LittleFS`, where you can select the size anywhere from 64KB to 512KB.


## Features

This library is just a simple `LittleFS` wrapper to facilitate your usage of `LittleFS` for the onboard flash on **MBED nRF52840-based boards such as Nano_33_BLE, Nano_33_BLE_Sense**, using [**Arduino-mbed mbed_nano** core](https://github.com/arduino/ArduinoCore-mbed)

The filesystem access uses normal [POSIX APIs](https://www.tutorialspoint.com/c_standard_library/stdio_h.htm) or [mbed FileSystem APIs](https://os.mbed.com/docs/mbed-os/v6.10/apis/file-system-apis.html)

---

### Currently supported Boards

1. **MBED nRF52840-based boards such as Nano_33_BLE, Nano_33_BLE_Sense**, etc. using [**Arduino-mbed mbed_nano** core](https://github.com/arduino/ArduinoCore-mbed)
2. **Seeeduino nRF52840-based boards such as SEEED_XIAO_NRF52840 and SEEED_XIAO_NRF52840_SENSE**, etc. using Seeeduino `mbed` core

---
---

## Prerequisites

1. [`Arduino IDE 1.8.19+` for Arduino](https://github.com/arduino/Arduino). [![GitHub release](https://img.shields.io/github/release/arduino/Arduino.svg)](https://github.com/arduino/Arduino/releases/latest)
2. [`Arduino mbed_nano core 3.4.1+`](https://github.com/arduino/ArduinoCore-mbed) for Arduino (Use Arduino Board Manager) MBED nRF52840-based boards such as **Nano_33_BLE, Nano_33_BLE_Sense**. [![GitHub release](https://img.shields.io/github/release/arduino/ArduinoCore-mbed.svg)](https://github.com/arduino/ArduinoCore-mbed/releases/latest)
3. `Seeeduino mbed core 2.7.2+` for Seeeduino nRF52840-based boards such as **SEEED_XIAO_NRF52840 and SEEED_XIAO_NRF52840_SENSE**

---
---

## Installation

### Use Arduino Library Manager

The best and easiest way is to use `Arduino Library Manager`. Search for [**FS_Nano33BLE**](https://github.com/khoih-prog/FS_Nano33BLE), then select / install the latest version.
You can also use this link [![arduino-library-badge](https://www.ardu-badge.com/badge/FS_Nano33BLE.svg?)](https://www.ardu-badge.com/FS_Nano33BLE) for more detailed instructions.

### Manual Install

Another way to install is to:

1. Navigate to [**FS_Nano33BLE**](https://github.com/khoih-prog/FS_Nano33BLE) page.
2. Download the latest release `FS_Nano33BLE-main.zip`.
3. Extract the zip file to `FS_Nano33BLE-main` directory 
4. Copy whole `FS_Nano33BLE-main` folder to Arduino libraries' directory such as `~/Arduino/libraries/`.

### VS Code & PlatformIO

1. Install [VS Code](https://code.visualstudio.com/)
2. Install [PlatformIO](https://platformio.org/platformio-ide)
3. Install [**FS_Nano33BLE** library](https://registry.platformio.org/libraries/khoih-prog/FS_Nano33BLE) by using [Library Manager](https://registry.platformio.org/libraries/khoih-prog/FS_Nano33BLE/installation). Search for **FS_Nano33BLE** in [Platform.io Author's Libraries](https://platformio.org/lib/search?query=author:%22Khoi%20Hoang%22)
4. Use included [platformio.ini](platformio/platformio.ini) file from examples to ensure that all dependent libraries will installed automatically. Please visit documentation for the other options and examples at [Project Configuration File](https://docs.platformio.org/page/projectconf.html)


---
---


### Examples 

 1. [FS_Counting](examples/FS_Counting)
 2. [FS_Test](examples/FS_Test)

---
---

### Example [FS_Test](examples/FS_Test)


https://github.com/khoih-prog/FS_Nano33BLE/blob/ac168d3003a8564ea0eb7a2e1ca7f4005f6f0f8f/examples/FS_Test/FS_Test.ino#L11-L371


---
---

### Debug Terminal Output Samples

### 1. FS_Counting on Nano 33 BLE with LittleFS size 256KB

The following is the sample terminal output when running example [FS_Counting](examples/FS_Counting) on MBED Nano_33_BLE using **LittleFS**

```cpp
Start FS_Test on Nano 33 BLE
LittleFS_Nano33BLE v1.2.1
FS_size (KB) = 256
FS_ Start Address = 0xC0000
[FS] LittleFS size (KB) = 256
[LFS] LittleFS Mount OK
Deleting file: /littlefs/counts.txt => OK
Times have been run = 1
 => Open to write OK

Start FS_Test on Nano 33 BLE
LittleFS_Nano33BLE v1.2.1
[LFS] LittleFS size (KB) = 256
[LFS] LittleFS Mount OK
 => Open to read OK
Times have been run = 2
 => Open to write OK

Start FS_Test on Nano 33 BLE
LittleFS_Nano33BLE v1.2.1
[LFS] LittleFS size (KB) = 256
[LFS] LittleFS Mount OK
 => Open to read OK
Times have been run = 3
 => Open to write OK
```

---

### 2. FS_Test on Nano 33 BLE with LittleFS size 256KB

The following is the sample terminal output when running example [FS_Test](examples/FS_Test) on MBED Nano_33_BLE using **LittleFS**


```cpp
Start FS_Test on Nano 33 BLE
LittleFS_Nano33BLE v1.2.1
FS_size (KB) = 256
FS_ Start Address = 0xC0000
[FS] LittleFS size (KB) = 256
[LFS] LittleFS Mount Fail
[LFS] Formatting... 
[LFS] 
FS Format OK. Mounting
====================================================
Writing file: /fs/hello1.txt => Open OK
* Writing OK
====================================================
Reading file: /fs/hello1.txt => Open OK
Hello from Nano_33_BLE
====================================================
Appending file: /fs/hello1.txt => Open OK
* Appending OK
====================================================
Reading file: /fs/hello1.txt => Open OK
Hello from Nano_33_BLE
Hello from Nano_33_BLE
====================================================
Renaming file: /fs/hello1.txt to: /fs/hello2.txt => OK
====================================================
readCharsFromFile: /fs/hello2.txt => Open OK
Hello from Nano_33_BLE
Hello from Nano_33_BLE
====================================================
Deleting file: /fs/hello2.txt => OK
====================================================
Reading file: /fs/hello2.txt => Open Failed
====================================================
Testing file I/O with: /fs/hello1.txt => Open OK
- writing

64 Kbytes written in (ms) 2461
====================================================
- reading

64 Kbytes read in (ms) 7
====================================================
Testing file I/O with: /fs/hello2.txt => Open OK
- writing

64 Kbytes written in (ms) 2460
====================================================
- reading

64 Kbytes read in (ms) 7
====================================================
Deleting file: /fs/hello1.txt => OK
====================================================
Deleting file: /fs/hello2.txt => OK
====================================================

Test complete
```

---

### 3. FS_Test on Nano 33 BLE with FATFS size 512KB

The following is the sample terminal output when running example [FS_Test](examples/FS_Test) on MBED Nano_33_BLE using **FATFS**


```cpp
Start FS_Test on Nano 33 BLE
FATFS_Nano33BLE v1.2.1
FS_size (KB) = 512
FS_ Start Address = 0x80000
[FS] LittleFS size (KB) = 512
[LFS] FATFS Mount OK
====================================================
Writing file: /fs/hello1.txt => Open OK
* Writing OK
====================================================
Reading file: /fs/hello1.txt => Open OK
Hello from Nano_33_BLE
====================================================
Appending file: /fs/hello1.txt => Open OK
* Appending OK
====================================================
Reading file: /fs/hello1.txt => Open OK
Hello from Nano_33_BLE
Hello from Nano_33_BLE
====================================================
Renaming file: /fs/hello1.txt to: /fs/hello2.txt => OK
====================================================
readCharsFromFile: /fs/hello2.txt => Open OK
Hello from Nano_33_BLE
Hello from Nano_33_BLE
====================================================
Deleting file: /fs/hello2.txt => OK
====================================================
Reading file: /fs/hello2.txt => Open Failed
====================================================
Testing file I/O with: /fs/hello1.txt => Open OK
- writing

64 Kbytes written in (ms) 4374
====================================================
- reading

64 Kbytes read in (ms) 15
====================================================
Testing file I/O with: /fs/hello2.txt => Open OK
- writing

64 Kbytes written in (ms) 4374
====================================================
- reading

64 Kbytes read in (ms) 15
====================================================
Deleting file: /fs/hello1.txt => OK
====================================================
Deleting file: /fs/hello2.txt => OK
====================================================

Test complete
```

---
---

### Debug

Debug is enabled by default on Serial.

You can also change the debugging level (_FS_LOGLEVEL_) from 0 to 4

```cpp
#define FS_DEBUG_OUTPUT    Serial

// These define's must be placed at the beginning before #include "FS_Nano33BLE.h"
// _FS_LOGLEVEL_ from 0 to 4
#define _FS_LOGLEVEL_      0
```

---

### Troubleshooting

If you get compilation errors, more often than not, you may need to install a newer version of the core for Arduino boards.

Sometimes, the library will only work if you update the board core to the latest version because I am using newly added functions.


---
---


### Issues

Submit issues to: [FS_Nano33BLE issues](https://github.com/khoih-prog/FS_Nano33BLE/issues)

---

## TO DO

1. Search for bug and improvement.


---

## DONE

1. Basic LittleFS wrapper for MBED nRF52840-based boards such as **Nano_33_BLE, Nano_33_BLE_Sense**, using [**Arduino-mbed mbed_nano** core](https://github.com/arduino/ArduinoCore-mbed)
2. Add Version String 
3. Add Table of Contents
4. Fix `multiple-definitions` linker error
5. Use correct `NANO33BLE_FS_START` address for `LittleFS` without wasting flash space. Check [Half size of flash #2](https://github.com/khoih-prog/FS_Nano33BLE/discussions/2)
6. Add support to Seeeduino nRF52840-based boards such as **SEEED_XIAO_NRF52840 and SEEED_XIAO_NRF52840_SENSE**, etc. using Seeeduino `mbed` core
7. Add astyle using `allman` style. Restyle the library
8. Display compile warning only when `_FS_LOGLEVEL_ > 3`


---
---

### Contributions and Thanks

Many thanks for everyone for bug reporting, new feature suggesting, testing and contributing to the development of this library.

1. Thanks to [Rob Probin](https://github.com/robzed) to report issue [Half size of flash #2](https://github.com/khoih-prog/FS_Nano33BLE/discussions/2) leading to v1.2.0

<table>
  <tr>
    <td align="center"><a href="https://github.com/robzed"><img src="https://github.com/robzed.png" width="100px;" alt="robzed"/><br /><sub><b>Rob Probin</b></sub></a><br /></td>
  </tr> 
</table>


---

## Contributing

If you want to contribute to this project:

- Report bugs and errors
- Ask for enhancements
- Create issues and pull requests
- Tell other people about this library

---

### License

- The library is licensed under [GPLv3](https://github.com/khoih-prog/FS_Nano33BLE/blob/main/LICENSE)

---

## Copyright

Copyright (c) 2021- Khoi Hoang




changelog.md
# FS_Nano33BLE Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/FS_Nano33BLE.svg?)](https://www.ardu-badge.com/FS_Nano33BLE)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/FS_Nano33BLE.svg)](https://github.com/khoih-prog/FS_Nano33BLE/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/FS_Nano33BLE/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/FS_Nano33BLE.svg)](http://github.com/khoih-prog/FS_Nano33BLE/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents

* [Changelog](#changelog)
  * [Releases v1.2.1](#releases-v121)
  * [Releases v1.2.0](#releases-v120)
  * [Releases v1.1.0](#releases-v110)
  * [Initial Releases v1.0.0](#initial-releases-v100)

---
---

## Changelog

### Releases v1.2.1

1. Add support to Seeeduino nRF52840-based boards such as **SEEED_XIAO_NRF52840 and SEEED_XIAO_NRF52840_SENSE**, etc. using Seeeduino `mbed` core
2. Add astyle using `allman` style. Restyle the library
3. Display compile warning only when `_FS_LOGLEVEL_ > 3`

### Releases v1.2.0

1. Use correct NANO33BLE_FS_START address without wasting flash space. Check [Half size of flash #2](https://github.com/khoih-prog/FS_Nano33BLE/discussions/2)
2. Enforce min 64KB / max 512KB of flash used for LittleFS by auto-adjusting NANO33BLE_FS_SIZE_KB if defined smaller than 64KB or larger than 512KB
3. Enforce 512KB of flash used for FATFS by auto-adjusting NANO33BLE_FS_SIZE_KB if defined different from 512KB
4. Update all examples

### Releases v1.1.0

1. Fix `multiple-definitions` linker error. Check [Different behaviour using the src_cpp or src_h lib #80](https://github.com/khoih-prog/ESPAsync_WiFiManager/discussions/80)
2. Update all examples

### Initial Releases v1.0.0

1. Initial coding to support MBED nRF52840-based boards such as **Nano_33_BLE, Nano_33_BLE_Sense**, etc. using [**Arduino-mbed mbed_nano** core](https://github.com/arduino/ArduinoCore-mbed)




src/FS_Nano33BLE.hpp
/****************************************************************************************************************************
  FS_Nano33BLE.hpp - Filesystem wrapper for FS (LittleFS and FATFS) on the Mbed Nano-33-BLE
  
  For MBED nRF52840-based boards such as Nano_33_BLE, Nano_33_BLE_Sense.
  Written by Khoi Hoang

  Built by Khoi Hoang https://github.com/khoih-prog/FS_Nano33BLE
  Licensed under MIT license

  Version: 1.0.0

  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  1.0.0   K Hoang      29/08/2021 Initial coding to support MBED nRF52840-based boards such as Nano_33_BLE, etc.
*****************************************************************************************************************************/

#ifndef _FS_NANO33BLE_HPP
#define _FS_NANO33BLE_HPP

#ifndef _FS_LOGLEVEL_
  #define _FS_LOGLEVEL_              1
#endif
 
bool FileSystem_MBED::init()
{
  FS_LOGERROR1(FS_NAME " size (KB) = ", NANO33BLE_FS_SIZE_KB);
  
#if FORCE_REFORMAT
  int error;
  
#if USING_LITTLEFS
  error = mbed::LittleFileSystem::format(&bd);
#else  
  error = mbed::FATFileSystem::format(&bd);
#endif

  if (error)
  {
    FS_LOGERROR1("FS Format error, errno = ", errno);
   
    return false;
  }

#endif

  return mount();
}

bool FileSystem_MBED::mount()
{
  if (!_mounted)
  {   
    int error;
    
    // 0 => OK
    _mounted = (fs.mount(&bd) == 0);

    FS_LOGERROR(_mounted ? FS_NAME " Mount OK" : FS_NAME " Mount Fail");

    if (!_mounted)
    {
      // Reformat if we can't mount the filesystem
      FS_LOGERROR("Formatting... ");

#if USING_LITTLEFS
      error = mbed::LittleFileSystem::format(&bd);
#else  
      error = mbed::FATFileSystem::format(&bd);
#endif

      if (!error)
      {
        FS_LOGERROR("FS Format OK. Mounting");

        _mounted = (fs.mount(&bd) == 0);
               
        if (_mounted)
        {
          FS_LOGERROR("FS OK");
          
          return true;
        }
        else
        {
          FS_LOGERROR("FS error. Reset. It'll be OK now after reset.");
          
          // Restart for nRF52
          NVIC_SystemReset();
          
          return false;
        }
      }
      else
      {
        FS_LOGERROR("FS Format error");
        
        return false;
      }
    }

    if (!_mounted)
    {
      FS_LOGERROR("FS error");
    } 
  }
  
  return true;
}

bool FileSystem_MBED::unmount()
{
  if (_mounted)
  {
    int err = fs.unmount();
  
    FS_LOGERROR1("Unmount FS ", err < 0 ? "Fail" : "OK");
 
    if (err < 0)
    {   
      return false;
    }
  }
  
  _mounted = false;
  
  return true;
}



#endif    // _FS_NANO33BLE_HPP




src/FS_Nano33BLE.h
/****************************************************************************************************************************
  FS_Nano33BLE.h - Filesystem wrapper for FS (LittleFS and FATFS) on the Mbed Nano-33-BLE

  For MBED nRF52840-based boards such as Nano_33_BLE, Nano_33_BLE_Sense.
  Written by Khoi Hoang

  Built by Khoi Hoang https://github.com/khoih-prog/FS_Nano33BLE
  Licensed under MIT license

  Version: 1.2.1

  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  1.0.0   K Hoang      29/08/2021 Initial coding to support MBED nRF52840-based boards such as Nano_33_BLE, etc.
  1.1.0   K Hoang      31/12/2021 Fix `multiple-definitions` linker error
  1.2.0   K Hoang      15/01/2022 Use correct NANO33BLE_FS_START address without wasting flash space
  1.2.1   K Hoang      25/10/2022 Add support to SEEED_XIAO_NRF52840 and SEEED_XIAO_NRF52840_SENSE using mbed
*****************************************************************************************************************************/

#ifndef _FS_NANO33BLE_H
#define _FS_NANO33BLE_H

#if !( defined(ARDUINO_ARCH_NRF52840) && defined(ARDUINO_ARCH_MBED) && \
     ( defined(ARDUINO_ARDUINO_NANO33BLE) || defined(ARDUINO_SEEED_XIAO_NRF52840) || defined(ARDUINO_SEEED_XIAO_NRF52840_SENSE) ) )
#error This code is intended to run on the MBED nRF52840 platform! Please check your Tools->Board setting.
#endif

#ifndef _FS_LOGLEVEL_
  #define _FS_LOGLEVEL_               1
#endif

////////////////////////////////////////////////

#ifndef USING_LITTLEFS
  #define USING_LITTLEFS              true
#endif

#if USING_LITTLEFS

  #define USING_FATFS                 false
  #define FS_NAME                     "LittleFS"

  #define MBED_FS_FILE_NAME           "littlefs"

  #if (_FS_LOGLEVEL_ > 3)
    #warning Use MBED nRF52840 (such as Nano_33_BLE, Nano_33_BLE_Sense) and LittleFS
  #endif

#else

  #define USING_FATFS                 true
  #define FS_NAME                     "FATFS"
  #define MBED_FS_FILE_NAME           "fs"

  #if (_FS_LOGLEVEL_ > 3)
    #warning Use MBED nRF52840 (such as Nano_33_BLE, Nano_33_BLE_Sense) and FATFS
  #endif

#endif

////////////////////////////////////////////////

#define FS_NANO33BLE_VERSION              FS_NAME "_Nano33BLE v1.2.1"

#define FS_NANO33BLE_VERSION_MAJOR        1
#define FS_NANO33BLE_VERSION_MINOR        2
#define FS_NANO33BLE_VERSION_PATCH        1

#define FS_NANO33BLE_VERSION_INT          1002001

////////////////////////////////////////////////

#include <Arduino.h>

#include "FS_Nano33BLE_Debug.h"

////////////////////////////////////////////////

// For Arduino' arduino-mbed core
// To check and determine if we need to init FS here
#if NANO33BLE_INITIALIZED
  #define MBED_FS_NEED_INIT               false

  #if (_FS_LOGLEVEL_ > 3)
    #warning NANO33BLE_INITIALIZED in another place
  #endif
#else
  // Better to delay until init done
  #if defined(NANO33BLE_INITIALIZED)
    #undef NANO33BLE_INITIALIZED
  #endif
  #define NANO33BLE_INITIALIZED           true

  #define MBED_FS_NEED_INIT               true

  #if (_FS_LOGLEVEL_ > 3)
    #warning NANO33BLE_INITIALIZED in FS_NANO33BLE
  #endif
#endif

#if MBED_FS_NEED_INIT

  //Use FS for MBED Nano_33_BLE
  #include "FlashIAPBlockDevice.h"

  #if USING_LITTLEFS
    #include "LittleFileSystem.h"
  #elif USING_FATFS
    #include "FATFileSystem.h"
  #else
    #error No FS selected
  #endif

  #include "mbed.h"

  #include <stdio.h>
  #include <errno.h>
  #include <functional>

  #include "BlockDevice.h"

  #if !defined(NANO33BLE_FLASH_SIZE)
    // Using max 512KB for FS
    #define NANO33BLE_FLASH_SIZE         (1 * 1024 * 1024)
  #endif

  #if !defined(NANO33BLE_FS_LOCATION_END)
    #define NANO33BLE_FS_LOCATION_END    NANO33BLE_FLASH_SIZE
  #endif

  #if !defined(NANO33BLE_FS_SIZE_KB)
    // Using default 64KB for FS
    #define NANO33BLE_FS_SIZE_KB       (64)

    #if (_FS_LOGLEVEL_ > 3)
      #warning Using default NANO33BLE_FS_SIZE_KB == 64KB
    #endif

  #else
    #if (_FS_LOGLEVEL_ > 3)
      #warning Using NANO33BLE_FS_SIZE_KB defined in external code
    #endif

    // KH, New from v1.2.0
    #if (NANO33BLE_FS_SIZE_KB > (NANO33BLE_FLASH_SIZE / (2 * 1024)))

      #if (_FS_LOGLEVEL_ > 3)
        #warning FlashSize too large. Adjust to 512KB
      #endif

      #undef NANO33BLE_FS_SIZE_KB
      #define NANO33BLE_FS_SIZE_KB        (512)

    #elif (NANO33BLE_FS_SIZE_KB < 64)

      #if (_FS_LOGLEVEL_ > 3)
        #warning FlashSize too small. Adjust to 64KB
      #endif

      #undef NANO33BLE_FS_SIZE_KB
      #define NANO33BLE_FS_SIZE_KB        (64)
    #endif
    //
  #endif

  #if !defined(NANO33BLE_FS_START)
    #define NANO33BLE_FS_START           (NANO33BLE_FLASH_SIZE - (NANO33BLE_FS_SIZE_KB * 1024))
  #endif

  #if !defined(FORCE_REFORMAT)
    #define FORCE_REFORMAT            false
  #elif FORCE_REFORMAT

    #if (_FS_LOGLEVEL_ > 3)
      #warning FORCE_REFORMAT enable. Are you sure ?
    #endif
  #endif

  #define MBED_FS_FILE_PREFIX         "/" MBED_FS_FILE_NAME

#endif    // MBED_FS_NEED_INIT

// nRF52840 flash address from 0, length 1MB.
// Better to use max half of flash for LitleFS, must be 0x80000 (512KB)
// FLASH_BASE must be 0x80000, or crash !!!????
#define FLASH_BASE            0x80000

#if USING_LITTLEFS
  static mbed::LittleFileSystem fs(MBED_FS_FILE_NAME);

  // KH, New from v1.2.0
  static FlashIAPBlockDevice bd(NANO33BLE_FS_START, (NANO33BLE_FS_SIZE_KB * 1024));
  //////
#elif USING_FATFS
  static mbed::FATFileSystem fs(MBED_FS_FILE_NAME);

  #if (NANO33BLE_FS_SIZE_KB != 512)

    #if (_FS_LOGLEVEL_ > 3)
      #warning Auto-adjust FATFS size to 512KB or crash
    #endif

    #undef NANO33BLE_FS_SIZE_KB
    #define NANO33BLE_FS_SIZE_KB        (512)
  #endif

  // KH, Note for v1.2.0
  // Don't use FATFS now or to use only with 512KB
  //static FlashIAPBlockDevice bd(NANO33BLE_FS_START, (NANO33BLE_FS_SIZE_KB * 1024));   // Still Crash, issue with the core ??
  //static FlashIAPBlockDevice bd(FLASH_BASE, (NANO33BLE_FS_SIZE_KB * 1024));           // Still Crash, issue with the core ??

  static FlashIAPBlockDevice bd(FLASH_BASE, (NANO33BLE_FS_SIZE_KB * 1024));
  //static FlashIAPBlockDevice bd(FLASH_BASE, NANO33BLE_FLASH_SIZE / 2);
  //////
#endif

class FileSystem_MBED
{
  public:
    FileSystem_MBED()
    {
      _size     = NANO33BLE_FS_SIZE_KB * 1024;
      _mounted  = false;
    }

    ~FileSystem_MBED()
    {
      if (_mounted)
      {
        unmount();
      }
    }

    ////////////////////////////////////////////////

    bool init()
    {
      FS_LOGERROR1(FS_NAME " size (KB) = ", NANO33BLE_FS_SIZE_KB);

#if FORCE_REFORMAT
      int error;

#if USING_LITTLEFS
      error = mbed::LittleFileSystem::format(&bd);
#else
      error = mbed::FATFileSystem::format(&bd);
#endif

      if (error)
      {
        FS_LOGERROR1("FS Format error, errno = ", errno);

        return false;
      }

#endif

      return mount();
    }

    ////////////////////////////////////////////////

    bool mount()
    {
      if (!_mounted)
      {
        int error;

        // 0 => OK
        _mounted = (fs.mount(&bd) == 0);

        FS_LOGERROR(_mounted ? FS_NAME " Mount OK" : FS_NAME " Mount Fail");

        if (!_mounted)
        {
          // Reformat if we can't mount the filesystem
          FS_LOGERROR("Formatting... ");

#if USING_LITTLEFS
          error = mbed::LittleFileSystem::format(&bd);
#else
          error = mbed::FATFileSystem::format(&bd);
#endif

          if (!error)
          {
            FS_LOGERROR("FS Format OK. Mounting");

            _mounted = (fs.mount(&bd) == 0);

            if (_mounted)
            {
              FS_LOGERROR("FS OK");

              return true;
            }
            else
            {
              FS_LOGERROR("FS error. Reset. It'll be OK now after reset.");

              // Restart for nRF52
              NVIC_SystemReset();

              return false;
            }
          }
          else
          {
            FS_LOGERROR("FS Format error");

            return false;
          }
        }

        if (!_mounted)
        {
          FS_LOGERROR("FS error");
        }
      }

      return true;
    }

    ////////////////////////////////////////////////

    bool unmount()
    {
      if (_mounted)
      {
        int err = fs.unmount();

        FS_LOGERROR1("Unmount FS ", err < 0 ? "Fail" : "OK");

        if (err < 0)
        {
          return false;
        }
      }

      _mounted = false;

      return true;
    }

    ////////////////////////////////////////////////

  private:

    uint32_t _size;

    bool     _mounted;
};


#endif // ifndef _FS_NANO33BLE_H
